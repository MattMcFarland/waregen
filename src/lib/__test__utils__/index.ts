import { clone, range } from "@@/utils";
import { DefaultWareEntity, DefaultBlueprintEntity } from "@@/entities/Ware";
import { AddwareEntity } from "@@/XMLTypes/X4WareGenXML";
import { GeneratorConfig } from "@@/generator";

export const mockGeneratorConfig = (): GeneratorConfig => ({
  defaultWare: defaultWare(),
  defaultBlueprint: defaultBlueprint(),
  addwaresList: mockAddwares(4),
  gamePath: "/test/Game",
  modPath: "extensions/test",
  modPrefix: "pfx",
  unpackedPath: "unpacked"
});

export const mockAddwares = (count: number): AddwareEntity[] => {
  const char = (code: number) => "abcdefghijklmnopqrstuvwxyz"[code];

  return range(count).map(
    (i: number): AddwareEntity => ({
      Attributes: {
        baskets: `[test_basket_${char(i)}]`,
        cloneProductionModuleFrom: `prod_cloned_macro_${char(i)}`,
        id: `${char(i)}`
      },
      ware: [mockWare(`${char(i)}`)],
      blueprint: [mockBlueprint(`${char(i)}`)]
    })
  );
};

export const mockWare = (name: string) => {
  const ware = defaultWare();
  ware.Attributes.name = name;
  return clone(ware);
};
export const mockBlueprint = (name: string) => {
  const bp = defaultBlueprint();
  bp.Attributes.name = `${name} production`;
  return clone(bp);
};

export const defaultBlueprint = (): DefaultBlueprintEntity => {
  const bp = <DefaultBlueprintEntity>defaultWare();
  bp.Attributes.tags = "module";
  bp.Attributes.description = "blueprint description";
  bp.restriction = [
    {
      Attributes: { licence: "station_gen_basic " }
    }
  ];
  bp.research = [
    {
      Attributes: { time: 10 },
      research: [
        { ware: [{ Attributes: { ware: "research_module_production" } }] }
      ]
    }
  ];
  bp.owner = [
    { Attributes: { faction: "antigone" } },
    { Attributes: { faction: "argon" } },
    { Attributes: { faction: "holyorder" } },
    { Attributes: { faction: "paranid" } },
    { Attributes: { faction: "teladi" } }
  ];
  bp.production = [
    {
      Attributes: {
        time: 791,
        amount: 1,
        method: "default",
        name: "{20206,101}"
      },
      primary: [
        {
          ware: [
            { Attributes: { ware: "claytronics", amount: 30 } },
            { Attributes: { ware: "energycells", amount: 75 } },
            { Attributes: { ware: "hullparts", amount: 150 } }
          ]
        }
      ]
    }
  ];
  return clone(bp);
};

export const defaultWare = (): DefaultWareEntity =>
  clone({
    Attributes: {
      id: "default",
      name: "default",
      transport: "container",
      volume: 1
    },
    price: [
      {
        Attributes: {
          min: 1,
          average: 1,
          max: 1
        }
      }
    ],
    production: [
      {
        Attributes: {
          time: 10,
          amount: 1,
          method: "default",
          name: "{20206, 101}"
        },
        primary: [
          {
            ware: [{ Attributes: { ware: "energycells", amount: 50 } }]
          }
        ],
        effects: [
          {
            effect: [
              {
                Attributes: { type: "efficiency", product: 1 }
              }
            ]
          }
        ]
      }
    ]
  });
