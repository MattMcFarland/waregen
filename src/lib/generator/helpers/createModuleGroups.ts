import { GeneratorConfig } from "..";
import { ModuleGroupsCollection, ModuleGroup } from "@@/entities";
import { IdRoster } from "@@/utils/IdRoster";

export default function createModuleGroups(
  config: GeneratorConfig
): ModuleGroupsCollection {
  const moduleGroups = config.addwaresList.map(addware => {
    const ids = new IdRoster(addware.Attributes.id, config.modPrefix);
    const group = new ModuleGroup();
    group.name = ids.production;
    group.macros = [ids.productionMacro];
    return group;
  });
  return new ModuleGroupsCollection(moduleGroups);
}
