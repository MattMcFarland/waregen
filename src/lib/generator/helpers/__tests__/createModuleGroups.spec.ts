import { mockGeneratorConfig } from "@@/__test__utils__";
import createModuleGroups from "../createModuleGroups";

describe("generator/helpers", () => {
  describe("createModuleGroups", () => {
    const mockConfig = mockGeneratorConfig();
    const modCollection = createModuleGroups(mockConfig);
    test("toJson()", () => {
      expect(modCollection.toJson()).toMatchSnapshot();
    });
    test("toXml()", () => {
      expect(modCollection.toXml()).toMatchSnapshot();
    });
  });
});
