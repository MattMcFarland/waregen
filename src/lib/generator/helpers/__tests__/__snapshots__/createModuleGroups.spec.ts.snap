// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generator/helpers createModuleGroups toJson() 1`] = `"{\\"group\\":[{\\"Attributes\\":{\\"name\\":\\"pfx_prod_gen_a\\"},\\"select\\":[{\\"Attributes\\":{\\"macro\\":\\"pfx_prod_gen_a_macro\\"}}]},{\\"Attributes\\":{\\"name\\":\\"pfx_prod_gen_b\\"},\\"select\\":[{\\"Attributes\\":{\\"macro\\":\\"pfx_prod_gen_b_macro\\"}}]},{\\"Attributes\\":{\\"name\\":\\"pfx_prod_gen_c\\"},\\"select\\":[{\\"Attributes\\":{\\"macro\\":\\"pfx_prod_gen_c_macro\\"}}]},{\\"Attributes\\":{\\"name\\":\\"pfx_prod_gen_d\\"},\\"select\\":[{\\"Attributes\\":{\\"macro\\":\\"pfx_prod_gen_d_macro\\"}}]},{\\"Attributes\\":{\\"name\\":\\"pfx_prod_gen_e\\"},\\"select\\":[{\\"Attributes\\":{\\"macro\\":\\"pfx_prod_gen_e_macro\\"}}]}]}"`;

exports[`generator/helpers createModuleGroups toXml() 1`] = `
"<groups>
  <group name=\\"pfx_prod_gen_a\\">
    <select macro=\\"pfx_prod_gen_a_macro\\"/>
  </group>
  <group name=\\"pfx_prod_gen_b\\">
    <select macro=\\"pfx_prod_gen_b_macro\\"/>
  </group>
  <group name=\\"pfx_prod_gen_c\\">
    <select macro=\\"pfx_prod_gen_c_macro\\"/>
  </group>
  <group name=\\"pfx_prod_gen_d\\">
    <select macro=\\"pfx_prod_gen_d_macro\\"/>
  </group>
  <group name=\\"pfx_prod_gen_e\\">
    <select macro=\\"pfx_prod_gen_e_macro\\"/>
  </group>
</groups>"
`;
